Fixing k8s 1.6 and 1.7 compatibility with CRD and TPR

aanm opened the pull request:
This PR was first created to deal with the huge SPAM of log messages stating `v1.NetworkPolicy does not exist` that were appearing in k8s `v1.6.x`. Instead, there were solved a couple of problems where the description and the solution for each one are resumed here.

**Problem 1**:

Since we dropped the support for `v1beta1.NetworkPolicy` in `0.10.0` we soon realized we should kept the old `v1beta1` for a one or 2 cilium versions (TBD). Without the support for `v1beta1` users running k8s `1.6.x` won't be able to test network policy enforcement with cilium.

**Solution** :

Cilium will detected if `v1.NetworkPolicy` is **not** available and will stop to listen for this object type. It will print a suggestion to a user running k8s `1.6.x` that he should update its kubernetes version to `>=1.7.x`

**Problem 2**:
Since we have upgrade k8s to `1.7.0` in the runtime tests, we started seeing failures that were occurring sporadically for the getting started guide for k8s and the L7 stresstests tests. We soon realized the kubernetes `1.7.0` had issues with Third Party Resources (TPR) and cilium wasn't receiving Cilium Network Policies from k8s api server, causing pods that weren't allow to talk with each other to talk with each other.

**Solution**:
Changed Cilium Network Policy definition to Custom Resource Definition (CRD). We changed all TPR code to CRD.

**Problem 3**:

CRD is not available in k8s <=`1.6.x`.

**Solution**:

Kept some of the TPR code so that we could have a transition period.

**Problem 4**:

TPR and CRD conflict if they are used at the same time in the same cluster. This causes Cilium to receive some updates of the cilium network policies added in the kubernetes cluster but if that policy is deleted then cilium might or might not receive that update.

**Solution**:

Bumped the CiliumNetworkPolicy definition to v2. If the user tries to run cilium in k8s `1.7.x` with TPR resources installed, he will see a warning message:

```
level=warning msg="Detected conflicting TPR and CRD, please migrate all ThirdPartyResource to CustomResourceDefinition! More info: https://cilium.link/migrate-tpr"
level=warning msg="Due conflicting TPR and CRD rules CiliumNetworkPolicy enforcement can't be guaranteed!"
```

The user will be in charge of changing all CiliumNetworkPolicy specs to v2.

TODO: ~Replace the short link with a cilium link~ giving some instructions regarding this issue with some examples on how to remove the old CiliumNetworkPolicy TPR.

**Problem 5**:

After all k8s issues were resolved we have noticed the L7 proxy library might have a small delay of 1-5 seconds if the system is under heavy load, when it's being set up to enforce L7 rules.

**Solution**:

Add a `sleep 10s` after setting up NetworkPolicies in kubernetes to give the L7 proxy more than reasonable time to set up.
Add a missing log.Error in case the L7 proxy fails to be set up.

The L7 proxy library will be fixed in a follow-up PR.

Fixes #947
Fixes #1168 
Fixes #1187

aanm committed:
k8s: removing k8s verbose v1.NetworkPolicy not found message

Since k8s 1.6 doesn't have support for the GA NetworkPolicy we print a
failure message if cilium tries to create a watcher for that resource.

To prevent the huge amount of error messages created by this issue we
simply stop the listener for v1.NetworkPolicy in k8s <1.7

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -62,17 +63,25 @@ func k8sErrorHandler(e error) {
+	if t, ok := k8sErrMsg[errstr]; !ok {
+		// Omitting the 'connection refused' common messages and
+		// Failed to list NetworkPolicy for k8s < 1.7
+		if strings.Contains(errstr, "connection refused") ||
```
This is better but we should check k8s version and omit it altogether if it has no chance of succeeding. I suggest we don't even start listening for NetworkPolicy for 1.7. Instead, in that case, a single warning message, in the beginning, should indicate that due to version, NetworkPolicy is not supported.

Once a minute still means  1440 duplicate log messages every day.

aanm committed:
k8s: re-adding v1beta1.NetworkPolicy

To ease the transaction between NetworkPolicy v1beta1 to GA this commit
re-adds the files used to parse v1beta1 NetworkPolicy spec files. These
files are meant to be kept until k8s 1.8.0.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: add v1beta1 network policy runtime tests

Signed-off-by: André Martins <andre@cilium.io>

Commit:
tests: fix logic to pick up k8s, cilium logs

* k8s tests: pick up kubelet, kube-controller-manager, kube-apiserver, cilium logs from both nodes.
* runtime tests: pick up cilium logs using journalctl
* name each tar.gz file using JOB_BASE_NAME-BUILD_NUMBER to differentiate between tar.gz files downloaded locally.

Signed-off by: Ian Vernon <ian@covalent.io>

aanm committed:
tests/k8s: adding policy get output

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: moving cilium network policies from TPR to CRD

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
vendor: adding dependency for CRD

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s/rbac: adding CRD permissions

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -94,24 +96,63 @@ func k8sErrorHandler(e error) {
 }
 
-func (d *Daemon) createThirdPartyResources() error {
+func createCustomResourceDefinitions(clientset apiextensionsclient.Interface) error {
```
This would mean that we can't use CNP < 1.7, correct?

aanm replied:
Correct. I can make ThirdPartyResource for k8s < 1.7, should I?

dqminh replied:
i was also working on some related changes for this. I think we should support both at least until tpr was deprecated fully. Probably something like this should suffice

```
ver, err := k8s.ClusterVersion()
useCRD := len(ver) >= 2 && ver[1] >= 7
// check and use crd, otherwise tpr
```

also we probably want to have a migrator that migrates things from tpr to crd ? not sure if its worth it though or users should just recreate things.

aanm replied:
Fixed

aanm committed:
k8s: re-adding TPR as fallback

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: testing multiple k8s versions

Enhanced k8s multi node tests to test cilium against multiple versions
of kubernetes.

For now it's testing against k8s v1.6.6 and v1.7.0

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -246,6 +260,51 @@ func (d *Daemon) deleteK8sNetworkPolicy(obj interface{}) {
+		return
+	}
+
+	log.Infof("Kubernetes v1beta1 network policy '%s' successfully add", k8sNP.Name)
```
nit: added

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -62,17 +64,28 @@ func k8sErrorHandler(e error) {
+			k8sErrMsg[errstr] = time.NewTimer(k8sErrLogTimeout)
+			k8sErrMsgMU.Unlock()
+		} else {
+			log.Warningf("Consider upgrading kubernetes to >=1.7 to enforce NetworkPolicy version 1")
```
Where is the `k8sErrMsgMU.Unlock()` in this path?

aanm replied:
I miss it, thanks!

aanm committed:
tests/k8s: added workarounds to test deprecated v1beta1 NP

In order to test on multiple k8s versions, each test needs to make sure
it doesn't try to install objects that don't exist on previous
kubernetes versions, e.g. NetworkPolicy v1 in k8s < 1.7

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests/k8s: enhanced cluster-manager flags

Since `yes-delete-all-etcd-data` flag was also removing containers and
removing kubeadm data, this commit splits up that option into 3:

- yes-delete-all-data
	deletes all kubeadm and etcd data
- yes-delete-kubeadm-data
	deletes containers and kubeadm relative data
- yes-delete-etcd-data
	deletes etcd data only

install_kubeadm function was also split into 2:

- install_kubeadm_dependencies
	installs kubeadm dependencies such as its repository and
	docker-engine
- install_kubeadm
	installs only kubeadmn

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Bumping jenkins timeout 60 minutes

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -62,17 +66,29 @@ func k8sErrorHandler(e error) {
+			k8sErrMsgMU.Unlock()
+		} else {
+			k8sErrMsgMU.Unlock()
+			log.Warningf("Consider upgrading kubernetes to >=1.7 to enforce NetworkPolicy version 1")
```
Shouldn't this log message be within the following if-block?

aanm replied:
Fixed!

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -81,19 +97,76 @@ func k8sErrorHandler(e error) {
+	}
+
+	log.Infof("k8s: Waiting for CRD to be established in k8s api-server...")
+	// wait for CRD being established
```
Wait for CRD to be established

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -81,19 +97,76 @@ func k8sErrorHandler(e error) {
+	if err != nil {
+		deleteErr := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(cnpCRDName, nil)
+		if deleteErr != nil {
+			return fmt.Errorf("k8s: unable to delete CRD %s. Deleting CRD due: %s", deleteErr, err)
```
due to:

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -110,20 +183,49 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
+	if err := createCustomResourceDefinitions(apiextensionsclientset); errors.IsNotFound(err) {
+		// If CRD was not found it might mean we are running in k8s <1.7
+		// then we should set up TPR instead
+		log.Debugf("Detected k8s <1.7, using TPR instead of CRD")
```
The log says it 'might mean' we are running k8s with version <1.7, but can we be sure of that? We should check the version of K8s via the API if possible, as this log statement might be incorrect based on the above comment.

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -81,19 +97,76 @@ func k8sErrorHandler(e error) {
+			Group:   k8s.CustomResourceDefinitionGroup,
+			Version: k8s.CustomResourceDefinitionVersion,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:     "ciliumnetworkpolicies",
```
Nit - would be better to have these strings as constants as some of them (e.g., the plural) are used elsewhere in the code.

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -246,6 +348,51 @@ func (d *Daemon) deleteK8sNetworkPolicy(obj interface{}) {
+	labels := labels.ParseSelectLabelArray(k8s.ExtractPolicyNameDeprecated(k8sNP))
+
+	if _, err := d.PolicyDelete(labels); err != nil {
+		log.Errorf("Error while deleting v1beta1 kubernetes network policy %+v: %s", labels, err)
```
General comment - a lot of these logs are inconsistent in how they phrase things, for example:
`v1beta1 kubernetes network policy` vs `kubernetes v1beta1`, `NetworkPolicy` vs `network policy`, `kubernetes` vs. `k8s`.

I think it's important for us to stick to convention when writing these logs, esp. for when we are searching logs to debug. If there is not a convention, we might miss out on key information.

ianvernon commented on pkg/k8s/network_policy_v1beta.go:
```diff
@@ -0,0 +1,132 @@
+			ingress.FromEndpoints = append(ingress.FromEndpoints, all)
+		} else {
+			for _, rule := range iRule.From {
+				// Only one or the other can be set, not both
```
This is saying that only rule.PodSelector or rule.NamespaceSelector can be set, but not both; should we check to see if both are set, and error out accordingly?

ianvernon commented on pkg/k8s/network_policy_v1beta.go:
```diff
@@ -0,0 +1,132 @@
+					if rule.PodSelector.MatchLabels == nil {
+						rule.PodSelector.MatchLabels = map[string]string{}
+					}
+					// The PodSelector should only reflect to the same namespace
```
'should only reflect the namespace for which the policy is being stored'

ianvernon commented on pkg/k8s/network_policy_v1beta.go:
```diff
@@ -0,0 +1,132 @@
+					rule.NamespaceSelector.MatchLabels = matchLabels
+
+					// We use our own special label prefix for namespace metadata,
+					// thus we need to prefix that prefix to all NamespaceSelector.MatchLabels
```
This comment is duplicated from above, and doesn't seem to apply here.

aanm committed:
daemon/k8s: removing TPR traces in a CRD cluster

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: printing l7 log file in case of failure

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
proxy: printing error in case of failure

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: adding netstat output to l7-stresstest

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: adding sleep for workaround

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: adding warning messages on a TPR and CRD cluster

If the user tries to set up k8s on a 1.7 cluster where TPRs are present
he will see the following warning messages:

```
Detected conflicting TPR and CRD, please migrate all ThirdPartyResource
to CustomResourceDefinition! More info: https://cilium.link/migrate-tpr

Due conflicting TPR and CRD rules CiliumNetworkPolicy enforcement can't
be guaranteed!
```

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests/k8s: updated all tests for the new CiliumNetworkPolicy v2

- CNP v1 will still be present in a suffixed `_deprecated` file so we
can run runtime tests against k8s 1.6.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests/k8s: Add workaround for kubeadm 1.7.0 boot error

When starting k8s 1.7.0 sometimes a 2nd node can't be added to the
cluster due a kubeadm race condition.

This presents a proposed workaround for that race that should be removed
once we upgrade kubeadm to 1.7.3.

Signed-off-by: André Martins <andre@cilium.io>
